### PR TITLE
Constraint programming extensions

### PR DESCRIPTION
**Pull request by Wikunia**
Using [ConstraintProgrammingExtensions.jl](https://github.com/dourouc05/ConstraintProgrammingExtensions.jl) 
for 
- [x] `AllDifferent`, 
- [x] `DifferentFrom`
- [x] `AllEqual`
- [x] Strictly still uses my bridge as it doesn't seem to be implemented in CPE
- [ ] TableSet is a matrix for me but a set of vectors in CPE. I think a matrix is easier for the user.
- [ ] BoolSets are fundamentally different in my solver so I'll stick with my implementation for now. See also https://github.com/dourouc05/ConstraintProgrammingExtensions.jl/issues/20

Ping @dourouc05 as you might be interested :wink:

### CODE REVIEW

File: docs/Manifest.toml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # This file is machine-generated - editing it directly is not advised
```
Comment (dourouc05) [inline docs/Manifest.toml:1]: I thought that Manifests should not be added to Git. (https://github.com/github/gitignore/blob/master/Julia.gitignore)

File: src/simplify.jl [hunk 2]
```diff
@@ -16,14 +16,14 @@ function simplify!(com)
     b_not_equal_to = false
```
Comment (dourouc05) [inline src/simplify.jl:16]: I'm wondering how much of this file could be useful more generally, for all CP solvers (something like https://github.com/mtanneau/MathOptPresolve.jl).

**Review by dourouc05 — APPROVED** [review]
For `Strictly`, my version has two parameters, but the constructors try to guess the right one: https://github.com/dourouc05/ConstraintProgrammingExtensions.jl/blob/master/src/sets_strictly.jl. Actually, you're using the same definition :): https://github.com/Wikunia/ConstraintSolver.jl/blob/master/src/types.jl#L171-L186. 

For `TableSet`, the order of the elements is not important, that's why I thought it was better to use a `Set` instead of a `Vector` (or a `Matrix`). The `Matrix` has the clear advantage to require that all elements have the same length, though (but that property can be enforced by the constructor). I've opened an issue to discuss this further: https://github.com/dourouc05/ConstraintProgrammingExtensions.jl/issues/21.

File: docs/Manifest.toml (hunk 1 above)
Comment (Wikunia) [reply docs/Manifest.toml:1]: It's the manifest file of the docs only which is quite helpful.

File: src/simplify.jl (hunk 2 above)
Comment (Wikunia) [reply src/simplify.jl:16]: That would definitely be useful in my opinion.

**Comment by Wikunia** [comment]
Haha I remember I copied Strictly from you. I probably just misread that. Yeah looks the same to me 😆

File: src/simplify.jl (hunk 2 above)
Comment (dourouc05) [reply src/simplify.jl:16]: We can think about it in greater detail later on, even MathOptPresolve is in its infancy.

File: docs/Manifest.toml (hunk 1 above)
Comment (Wikunia) [reply docs/Manifest.toml:1]: Not needed though I reckon

**Comment by dourouc05** [comment]
I thought that this kind of commit would reduce the amount of code in the solver, but you're adding 50 lines with it :/. Do you think anything should be done better in CPE? (While remaining at MOI level, without dependency on JuMP, that is.)

File: docs/Manifest.toml (hunk 1 above)
Comment (dourouc05) [reply docs/Manifest.toml:1]: I frankly have no opinion on this.

**Comment by Wikunia** [comment]
The extra lines are probably from the manifest file. Otherwise I only removed lines and added your package

**Comment by odow** [comment]
Great to see this coming along.

**Comment by dourouc05** [comment]
For users, maybe it would be best if you provided `@deprecate` replacements for building sets? This would print a warning to change the calls, which could make updating their version of your solver easier. Normally, the syntax is: `@deprecate old new false`, without parentheses (I just hope that it works across packages :/).

**Comment by Wikunia** [comment]
I'm actually wondering whether I should export `ConstraintSolverExtensions.AllDifferent` and others, such that the user can use `AllDifferent` instead of needing to have `using ConstraintSolverExtensions` as well. What do you think in general about how solvers should handle this @dourouc05 ?

**Comment by dourouc05** [comment]
Based on what MOI solvers have, they do not export anything: rather, every set is exported by JuMP if need be (and the symbol `MOI` is exported by JuMP too, so that you have direct access for sets that do not have a shortcut in JuMP). In my opinion, it's the job of JuCP (or whatever it ends up being called) to do that; in the meantime, it's probably OK that solvers do it too.

**Comment by dourouc05** [comment]
Indeed, I don't have yet bridges to reverse inequalities. Do you want to make a PR with https://github.com/Wikunia/ConstraintSolver.jl/blob/master/src/MOI_wrapper/Bridges/strictly_greater_than.jl or can I copy that code?

**Comment by Wikunia** [comment]
Feel free to copy the code :wink:

**Comment by dourouc05** [comment]
Yay :D! I just added a link to this solver from CPE's README.

**Comment by dourouc05** [comment]
If you want, you can replace your bridge by the new one: https://github.com/dourouc05/ConstraintProgrammingExtensions.jl/commit/660a0d6225d5a33f2a977e57ff97e8232cbd0637. I don't think your other bridges make sense in CPE for the moment, as long as we haven't settled on how to represent Boolean constraints.

**Pull request merged**